File: server/src/main/scala/kpn/api/custom/Tags.scala
package kpn.api.custom

object Tags {
  private val ValueSeparator = ";"
  type TagTuple = (String, String)

  def from(tags: TagTuple*): Seq[Tag] = {
    tags.map { case (key, value) => Tag(key, value) }
  }

  def from(map: Map[String, String]): Seq[Tag] = {
    map.toSeq.map { case (key, value) => Tag(key, value) }
  }

  def toString(tags: Seq[Tag]): String = tags.map { case Tag(key, value) => s"$key=$value" }.mkString(", ")

  def get(tags: Seq[Tag], key: String): Option[String] = {
    tags.find(_.key == key).map(_.value)
  }

  def values(tags: Seq[Tag], key: String): Seq[String] = {
    get(tags, key)
      .filter(_.nonEmpty)
      .map(splitAndNormalize)
      .getOrElse(Seq.empty)
  }

  def has(tags: Seq[Tag], key: String, allowedValues: String*): Boolean = {
    tags.exists { tag =>
      tag.key == key && (
        allowedValues.isEmpty ||
          splitAndNormalize(tag.value).exists(allowedValues.contains)
        )
    }
  }

  def splitAndNormalize(value: String): Seq[String] = {
    if (value.contains(ValueSeparator)) {
      value.split(ValueSeparator)
        .toSeq
        .map(_.trim)
        .filter(_.nonEmpty)
        .sorted
    } else {
      Seq(value.trim)
    }
  }
}

File: backend/api/src/main/java/kpn/api/custom/Tags.java
package kpn.api.custom;

import kpn.api.custom.Tag;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;

public final class Tags {

  private static final String VALUE_SEPARATOR = ";";

  private Tags() {
  }

  @SafeVarargs
  public static List<Tag> from(Map.Entry<String, String>... tags) {
    return Arrays.stream(tags)
      .map(entry -> new Tag(entry.getKey(), entry.getValue()))
      .toList();
  }

  public static List<Tag> from(ImmutableMap<String, String> map) {
    return map.entrySet().stream()
      .map(entry -> new Tag(entry.getKey(), entry.getValue()))
      .toList();
  }

  public static String toString(List<Tag> tags) {
    return tags.stream()
      .map(tag -> tag.key() + "=" + tag.value())
      .collect(Collectors.joining(", "));
  }

  public static Optional<String> get(List<Tag> tags, String key) {
    return tags.stream()
      .filter(tag -> tag.key().equals(key))
      .map(Tag::value)
      .findFirst();
  }

  public static ImmutableList<String> values(List<Tag> tags, String key) {
    return get(tags, key)
      .filter(value -> !value.isEmpty())
      .map(kpn.api.custom.Tags::splitAndNormalize)
      .orElse(ImmutableList.of());
  }

  public static boolean has(List<Tag> tags, String key, String... allowedValues) {
    return tags.stream().anyMatch(tag ->
      tag.key().equals(key) && (
        allowedValues.length == 0 ||
          splitAndNormalize(tag.value()).stream()
            .anyMatch(value -> Arrays.asList(allowedValues).contains(value))
      )
    );
  }

  public static ImmutableList<String> splitAndNormalize(String value) {
    if (value.contains(VALUE_SEPARATOR)) {
      return Arrays.stream(value.split(VALUE_SEPARATOR))
        .map(String::trim)
        .filter(part -> !part.isEmpty())
        .sorted()
        .collect(ImmutableList.toImmutableList());
    } else {
      return ImmutableList.of(value.trim());
    }
  }
}
